managed/services/telemetry/telemetry.go

```diff
@@ -160,7 +161,61 @@ func (s *Service) DistributionMethod() serverpb.DistributionMethod {
 	return s.sDistributionMethod
 }
 
-func (s *Service) prepareReport(ctx context.Context) *reporter.ReportRequest {
+func (s *Service) processSendCh(ctx context.Context) {
+	var reportsSync sync.Mutex
+	var inflightReports []*pmmv1.ServerMetric
+	var sendCtx context.Context
+	var cancel context.CancelFunc
+
+	cleanup := func() {
+		inflightReports = nil
+		sendCtx = nil
+		cancel = nil
+	}
+
+	for {
+		select {
+		case report, ok := <-s.sendCh:
+			if ok {
+				s.l.Debug("Processing telemetry report.")
+				if sendCtx != nil {
+					cancel()
+				}
+				reportsSync.Lock()
+				inflightReports = append(inflightReports, report)
+				sendCtx, cancel = context.WithCancel(ctx)
+
+				reportsCopy := make([]*pmmv1.ServerMetric, len(inflightReports))
+				copy(reportsCopy, inflightReports)
+				reportsSync.Unlock()
+
+				go func(ctx context.Context, reports []*pmmv1.ServerMetric, cleanup func()) {
+					reportsSync.Lock()
+					defer reportsSync.Unlock()
```

Comment:
now there is no sense in moving this functionality into a separate go routine since the mutex is locked and L184 will wait while the sending is finished -> usual sync sending

Reply:
Don't agree, request will be canceled in other goroutine, this part is out of critical section